File: sonar-squid/src/main/java/org/sonar/squid/text/Line.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.sonar.squid.text;

import org.apache.commons.lang.StringUtils;
import org.sonar.squid.measures.Measurable;
import org.sonar.squid.measures.Metric;

class Line implements Measurable<Metric> {

  private static final String NOSONAR_TAG = "NOSONAR";

  private final int lineIndex;
  private int blankLine = 0;
  private int line = 1;
  private int lineOfCode = 0;
  private int commentLine = 0;
  private int headerCommentLine = 0;
  private int commentBlankLine = 0;
  private int commentedOutCodeLine = 0;
  private String comment = null;
  private StringBuilder stringLine;
  private boolean isBlank;
  private boolean isThereJavadoc;
  private boolean isThereLicenseHeaderComment;

  Line() {
    this.lineIndex = 0;
  }

  Line(String stringLine) {
    this();
    setString(new StringBuilder(stringLine));
  }

  Line(int lineIndex, StringBuilder stringLine) {
    this(lineIndex);
    setString(stringLine);
  }

  Line(int lineIndex) {
    this.lineIndex = lineIndex;
  }

  final void setString(StringBuilder stringLine) {
    this.stringLine = stringLine;
    isBlank = isBlankLine();
  }

  private boolean isBlankLine() {
    for (int i = 0; i < stringLine.length(); i++) {
      if ( !Character.isWhitespace(stringLine.charAt(i))) {
        return false;
      }
    }
    return true;
  }

  /**
   * {@inheritDoc}
   */
  public double getDouble(Metric metric) {
    return getInt(metric);
  }

  /**
   * {@inheritDoc}
   */
  public int getInt(Metric metric) {
    switch (metric) {
      case BLANK_LINES:
        return blankLine;
      case LINES:
        return line;
      case LINES_OF_CODE:
        return lineOfCode;
      case COMMENT_LINES:
        return commentLine;
      case COMMENTED_OUT_CODE_LINES:
        return commentedOutCodeLine;
      case COMMENT_BLANK_LINES:
        return commentBlankLine;
      case HEADER_COMMENT_LINES:
        return headerCommentLine;
      default:
        throw new IllegalStateException("Metric " + metric.name() + " is not available on Line object.");
    }
  }

  /**
   * {@inheritDoc}
   */
  public void setMeasure(Metric metric, double measure) {
    setMeasure(metric, (int) measure);
  }

  /**
   * {@inheritDoc}
   */
  public void setMeasure(Metric metric, int measure) {
    switch (metric) {
      case BLANK_LINES:
        blankLine = measure;
        break;
      case LINES_OF_CODE:
        lineOfCode = measure;
        break;
      case COMMENT_LINES:
        commentLine = measure;
        break;
      case COMMENTED_OUT_CODE_LINES:
        commentedOutCodeLine = measure;
        break;
      case COMMENT_BLANK_LINES:
        commentBlankLine = measure;
        break;
      case HEADER_COMMENT_LINES:
        headerCommentLine = measure;
        break;
      case LINES:
        throw new IllegalStateException("Metric LINES always equals 1 on a Line and you are not permitted to change this value.");
      default:
        throw new IllegalStateException("Metric " + metric.name() + " is not suitable for Line object.");
    }
  }

  void setComment(String comment) {
    this.comment = comment;
  }

  void setComment(String comment, boolean isJavadoc) {
    setComment(comment);
    this.isThereJavadoc = isJavadoc;
  }

  void setComment(String comment, boolean isJavadoc, boolean isLicenseHeader) {
    setComment(comment, isJavadoc);
    this.isThereLicenseHeaderComment = isLicenseHeader;
  }

  String getString() {
    return stringLine.toString();
  }

  boolean isBlank() {
    return !isThereComment() && isBlank;
  }

  boolean isThereCode() {
    if ( !isBlank() && !isThereComment()) {
      return true;
    }
    if (isThereComment() && isThereCodeBeforeOrAfterComment()) {
      return true;
    }
    return false;
  }

  private boolean isThereCodeBeforeOrAfterComment() {
    if ( !isThereComment()) {
      throw new IllegalStateException("You can't call this method when there isn't any comment");
    }
    boolean isThereCodeBeforeComment = false;
    boolean isThereCodeAfterComment = false;
    int commentStartIndex = stringLine.indexOf(comment);
    int commentEndIndex = commentStartIndex + comment.length() - 1;
    if (commentStartIndex > 0) {
      isThereCodeBeforeComment = !StringUtils.isBlank(stringLine.substring(0, commentStartIndex - 1));
    }
    if (commentEndIndex > 0 && commentEndIndex != stringLine.length() - 1) {
      isThereCodeAfterComment = !StringUtils.isBlank(stringLine.substring(commentEndIndex + 1));
    }
    return isThereCodeBeforeComment || isThereCodeAfterComment;
  }

  boolean isThereComment() {
    return comment != null;
  }

  String getComment() {
    return comment;
  }

  boolean isThereBlankComment() {
    if (isThereComment()) {
      for (int i = 0; i < comment.length(); i++) {
        char character = comment.charAt(i);
        if ( !Character.isWhitespace(character) && character != '*' && character != '/') {
          return false;
        }
      }
      return true;
    }
    return false;
  }

  boolean isThereJavadoc() {
    return isThereJavadoc;
  }

  boolean isThereLicenseHeaderComment() {
    return isThereLicenseHeaderComment;
  }

  boolean isThereNoSonarTag() {
    return isThereComment() && comment.contains(NOSONAR_TAG);
  }

  int getLineIndex() {
    return lineIndex;
  }

  void deleteLineContent() {
    comment = null;
    stringLine = null;
  }
}

File: server/sonar-web/src/main/js/components/SourceViewer/components/Line.js
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import classNames from 'classnames';
import { times } from 'lodash';
import LineNumber from './LineNumber';
import LineSCM from './LineSCM';
import LineCoverage from './LineCoverage';
import LineDuplications from './LineDuplications';
import LineDuplicationBlock from './LineDuplicationBlock';
import LineIssuesIndicator from './LineIssuesIndicator';
import LineCode from './LineCode';
/*:: import type { SourceLine } from '../types'; */
/*:: import type { LinearIssueLocation } from '../helpers/indexing'; */
/*:: import type { Issue } from '../../issue/types'; */

/*::
type Props = {|
  branch?: string,
  displayAllIssues: boolean,
  displayCoverage: boolean,
  displayDuplications: boolean,
  displayIssues: boolean,
  displayIssueLocationsCount?: boolean;
  displayIssueLocationsLink?: boolean;
  displayLocationMarkers?: boolean;
  duplications: Array<number>,
  duplicationsCount: number,
  filtered: boolean | null,
  highlighted: boolean,
  highlightedLocationMessage?: { index: number, text: string },
  highlightedSymbols?: Array<string>,
  issueLocations: Array<LinearIssueLocation>,
  issues: Array<Issue>,
  last: boolean,
  line: SourceLine,
  loadDuplications: SourceLine => void,
  onClick: (SourceLine, HTMLElement) => void,
  onCoverageClick: (SourceLine, HTMLElement) => void,
  onDuplicationClick: (number, number) => void,
  onIssueChange: Issue => void,
  onIssueSelect: string => void,
  onIssueUnselect: () => void,
  onIssuesOpen: SourceLine => void,
  onIssuesClose: SourceLine => void,
  onLocationSelect?: number => void,
  onSCMClick: (SourceLine, HTMLElement) => void,
  onSymbolClick: (Array<string>) => void,
  openIssues: boolean,
  onPopupToggle: (issue: string, popupName: string, open: ?boolean ) => void,
  openPopup: ?{ issue: string, name: string},
  previousLine?: SourceLine,
  scroll?: HTMLElement => void,
  secondaryIssueLocations: Array<{
    from: number,
    to: number,
    line: number,
    index: number,
    startLine: number
  }>,
  selectedIssue: string | null
|};
*/

export default class Line extends React.PureComponent {
  /*:: props: Props; */

  handleIssuesIndicatorClick = () => {
    if (this.props.openIssues) {
      this.props.onIssuesClose(this.props.line);
      this.props.onIssueUnselect();
    } else {
      this.props.onIssuesOpen(this.props.line);

      const { issues } = this.props;
      if (issues.length > 0) {
        this.props.onIssueSelect(issues[0].key);
      }
    }
  };

  render() {
    const { line, duplications, displayCoverage, duplicationsCount, filtered } = this.props;
    const className = classNames('source-line', {
      'source-line-highlighted': this.props.highlighted,
      'source-line-filtered': filtered === true,
      'source-line-filtered-dark':
        displayCoverage &&
        (line.coverageStatus === 'uncovered' || line.coverageStatus === 'partially-covered'),
      'source-line-last': this.props.last
    });

    return (
      <tr className={className} data-line-number={line.line}>
        <LineNumber line={line} onClick={this.props.onClick} />

        <LineSCM
          line={line}
          onClick={this.props.onSCMClick}
          previousLine={this.props.previousLine}
        />

        {this.props.displayCoverage && (
          <LineCoverage line={line} onClick={this.props.onCoverageClick} />
        )}

        {this.props.displayDuplications && (
          <LineDuplications line={line} onClick={this.props.loadDuplications} />
        )}

        {times(duplicationsCount).map(index => (
          <LineDuplicationBlock
            duplicated={duplications.includes(index)}
            index={index}
            key={index}
            line={this.props.line}
            onClick={this.props.onDuplicationClick}
          />
        ))}

        {this.props.displayIssues &&
          !this.props.displayAllIssues && (
            <LineIssuesIndicator
              issues={this.props.issues}
              line={line}
              onClick={this.handleIssuesIndicatorClick}
            />
          )}

        <LineCode
          branch={this.props.branch}
          displayIssueLocationsCount={this.props.displayIssueLocationsCount}
          displayIssueLocationsLink={this.props.displayIssueLocationsLink}
          displayLocationMarkers={this.props.displayLocationMarkers}
          highlightedLocationMessage={this.props.highlightedLocationMessage}
          highlightedSymbols={this.props.highlightedSymbols}
          issues={this.props.issues}
          issueLocations={this.props.issueLocations}
          line={line}
          onIssueChange={this.props.onIssueChange}
          onIssueSelect={this.props.onIssueSelect}
          onLocationSelect={this.props.onLocationSelect}
          onSymbolClick={this.props.onSymbolClick}
          onPopupToggle={this.props.onPopupToggle}
          openPopup={this.props.openPopup}
          scroll={this.props.scroll}
          secondaryIssueLocations={this.props.secondaryIssueLocations}
          selectedIssue={this.props.selectedIssue}
          showIssues={this.props.openIssues || this.props.displayAllIssues}
        />
      </tr>
    );
  }
}
